modyn/supervisor/internal/grpc_handler.py

```diff
@@ -434,3 +462,173 @@ def seed_selector(self, seed: int) -> None:
+            if "label_transformer_function" in dataset:
+                label_transformer = dataset["label_transformer_function"]
+            else:
+                label_transformer = ""
+
+            bytes_parser_function = dataset["bytes_parser_function"]
+            batch_size = dataset["batch_size"]
+            dataloader_workers = dataset["dataloader_workers"]
+            metrics = []
+            for metric in dataset["metrics"]:
+                name = metric["name"]
+                if "config" in metric:
+                    metric_config = json.dumps(metric["config"])
+                else:
+                    metric_config = "{}"
+
+                if "evaluation_transformer_function" in metric:
+                    evaluation_transformer = metric["evaluation_transformer_function"]
+                else:
+                    evaluation_transformer = ""
+
+                metrics.append(
+                    MetricConfiguration(
+                        name=name,
+                        config=EvaluatorJsonString(value=metric_config),
+                        evaluation_transformer=EvaluatorPythonString(value=evaluation_transformer),
+                    )
+                )
+
+            start_evaluation_kwargs = {
+                "trained_model_id": trained_model_id,
+                "dataset_info": DatasetInfo(dataset_id=dataset_id, num_dataloaders=dataloader_workers),
+                "device": device,
+                "amp": amp,
+                "batch_size": batch_size,
+                "metrics": metrics,
+                "model_id": model_id,
+                "model_configuration": EvaluatorJsonString(value=model_config),
+                "transform_list": transform_list,
+                "bytes_parser": EvaluatorPythonString(value=bytes_parser_function),
+                "label_transformer": EvaluatorPythonString(value=label_transformer),
+            }
+
+            cleaned_kwargs = {k: v for k, v in start_evaluation_kwargs.items() if v is not None}
+
+            req = EvaluateModelRequest(**cleaned_kwargs)
+
+            response: EvaluateModelResponse = self.evaluator.evaluate_model(req)
+
+            if not response.evaluation_started:
+                logger.error(f"Starting evaluation for dataset {dataset_id} did go wrong: {response}")
+            else:
+                evaluation_id = response.evaluation_id
+                logger.info(f"Started evaluation {evaluation_id} on dataset {dataset_id}.")
+                evaluations[evaluation_id] = EvaluationStatusTracker(dataset_id, response.dataset_size)
+
+        return evaluations
+
+    def wait_for_evaluation_completion(self, training_id: int, evaluations: dict[int, EvaluationStatusTracker]) -> None:
+        if not self.connected_to_evaluator:
+            raise ConnectionError("Tried to wait for evaluation to finish, but not there is no gRPC connection.")
+
+        self.status_bar.update(demo=f"Waiting for evaluation (training = {training_id})")
+
```

Comment:
Could you maybe explain what is happening here? I am not sure why we need a deque here (compared to the wait training to finish function), and if we can somehow make a common wait function out of these since they are very similar. If you could add some comments/explain the ideas here that would be great, right now it is hard for me to understand the idea (especially coming from the training thing without a deque)

Reply:
Sure! Compared to training, here we have a parallel setup. The evaluator might receive several evaluation requests (for different evaluation datasets) and works on them in parallel. Hence, we also fetch the status in the supervisor for all ongoing evaluations in this `wait_for_evaluation_completion` function. I used a dequeue here in order to fetch the status of all evaluations sequentially in a round-robin manner. (We can also think of adding a thread for each evaluation at this point, which might simplify the unification process for waiting in training and evaluation).

Reply:
Speaking of a common wait function: We introduced optional attributes in the `trainer_server.proto` (see `TrainingStatusResponse`) in #256. Was there a particular reason doing this?

Reply:
We had some discussions on having a field set to 0 + a success flag vs optional. I don't really remember any other reason than potential clarity in the code I think. Is that a problem?

Reply:
I see, thank you! No, this is not a problem. I've used the same approach for evaluation. Note, however, that I did not make a common waiting function. They are indeed (very) similar but are using other classes for e.g status bar, training has downsampling and evaluation handles several `EvaluationStatusRequest` in parallel.

Comment done.